### PR TITLE
✨Add form support for Linker

### PR DESCRIPTION
**Pull request by calebcordry**
Enable Linker to pass data on form submit. This is done by adding a `<input>` tag to all forms containing any linker data.

### CODE REVIEW

File: extensions/amp-analytics/0.1/utils.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Returns an string with a total of 128 of random values based on the
+ * `win.crypto.getRandomValues` API. If that is not available concatenates
+ * a string of other values that might be hard to guess including
+ * `Math.random` and the current time.
+ * @param {!Window} win
+ * @return {string} Entropy.
+ */
+export function getUniqueId(win) {
```
Comment (lannka) [inline extensions/amp-analytics/0.1/utils.js]: this is overkill, let's just use a counter.

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 2]
```diff
@@ -94,6 +102,8 @@ export class LinkerManager {
         });
         this.resolvedLinkers_[name] =
             createLinker(/* version */ '1', expandedIds);
+        // This resolution indicates that we have finished a single linker.
+        return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: Done.

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 3]
```diff
@@ -255,8 +269,69 @@ export class LinkerManager {
...
+    // Wait for all linker macros to finish resolving.
+    linkersFinalPromise.then(() => {
+      const doc = this.ampdoc_.getRootNode();
+
+      // Maybe a new form has been added.
+      doc.addEventListener(AmpEvents.DOM_UPDATE, () => {
+        this.maybeAddDataToForms_(doc.querySelectorAll('form'));
+      });
+
+      const forms = doc.querySelectorAll('form');
```
[the remark this answers is not in the document]
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: Done.

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 4]
```diff
@@ -103,7 +113,10 @@ export class LinkerManager {
           this.handleAnchorMutation_.bind(this), Priority.ANALYTICS_LINKER);
     }
 
-    return Promise.all(this.allLinkerPromises_);
+    const linkersFinalPromise = Promise.all(this.allLinkerPromises_);
+    this.enableFormSupport_(linkersFinalPromise);
+
+    return linkersFinalPromise;
```
[the remark this answers is not in the document]
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: Done.

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 5]
```diff
@@ -255,8 +269,69 @@ export class LinkerManager {
...
+  }
+
+  /**
+   * Check to see if we have already added the linker kv pairs to this form,
+   * if not, add them.
+   * @param {!NodeList<!Element>} forms
+   */
+  maybeAddDataToForms_(forms) {
+    iterateCursor(forms, form => {
+      if (form.linkerIds_ && form.linkerIds_[this.uid_]) {
```
[the remark this answers is not in the document]
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: Done.

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 6]
```diff
@@ -255,8 +269,69 @@ export class LinkerManager {
...
+   */
+  addDataToForm_(form) {
+    Object.keys(this.resolvedLinkers_).forEach(key => {
+      const attrs = dict({
+        'type': 'hidden',
+        'name': key,
+        'value': this.resolvedLinkers_[key],
+      });
+      const inputEl = createElementWithAttributes(
+          /** @type {!Document} */ (this.ampdoc_.getRootNode()),
```
[the remark this answers is not in the document]
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: changed to `form.ownerDocument.

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 7]
```diff
@@ -255,8 +269,69 @@ export class LinkerManager {
 
     el.href = addParamToUrl(href, name, this.resolvedLinkers_[name]);
   }
+
+  /**
+   * Add data to any existing forms, and listen for any new forms that may be
+   * created.
+   * @param {Promise} linkersFinalPromise
+   */
+  enableFormSupport_(linkersFinalPromise) {
```
[the remark this answers is not in the document]
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: Note to myself: this is not yet fixed (and was a silly oversight 😅)

File: extensions/amp-analytics/0.1/utils.js (hunk 1 above)
Comment (calebcordry) [reply extensions/amp-analytics/0.1/utils.js]: Done.

File: extensions/amp-analytics/0.1/linker-manager.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: Done 👍

**Comment by calebcordry** [comment]
Large refactor to use form submit service complete 🙂

File: extensions/amp-analytics/0.1/linker-manager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (lannka) [reply extensions/amp-analytics/0.1/linker-manager.js]: that does seem to be right. L92 does return a promise.

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 8]
```diff
@@ -230,15 +239,27 @@ export class LinkerManager {
...
+
+    if (this.isDomainMatch_(hostname, domains)) {
+      el.href = addParamToUrl(href, name, this.resolvedLinkers_[name]);
+    }
+  }
+
+  /**
+   * Check to see if the url is a match for the given set of domains.
+   * @param {string} url
+   * @param {Array} domains
```
Comment (lannka) [inline extensions/amp-analytics/0.1/linker-manager.js]: {!Array}

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 9]
```diff
@@ -230,15 +239,27 @@ export class LinkerManager {
...
+   */
+  isDomainMatch_(url, domains) {
```
Comment (lannka) [inline extensions/amp-analytics/0.1/linker-manager.js]: hostname

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 10]
```diff
@@ -247,13 +268,62 @@ export class LinkerManager {
...
+        return false;
       }
     }
+    return true;
+  }
+
+  /**
+   * Register callback that will handle form sumbits.
+   */
+  enableFormSupport_() {
```
Comment (lannka) [inline extensions/amp-analytics/0.1/linker-manager.js]: we will need to remove the listener when amp-analytics unlayout. (can be a separate PR).

File: extensions/amp-analytics/0.1/linker-manager.js (hunk 8 above)
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: `{?Array}` Done.

File: extensions/amp-analytics/0.1/linker-manager.js (hunk 9 above)
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: Done.

File: extensions/amp-analytics/0.1/linker-manager.js (hunk 10 above)
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: SGTM will make follow-up PR

File: extensions/amp-analytics/0.1/linker-manager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: You are right (as usual). I tried testing this before but must have been doing something wrong 🤷‍♂️

File: src/services.js [hunk 11]
```diff
@@ -225,15 +225,28 @@ export class Services {
   }
 
   /**
-   * Returns service implemented in service/history-impl.
+   * Returns a service to register callbacks we wish to execute when an
+   * amp-form is submitted. This is the sync version used by amp-form only, all
```
Comment (lannka) [inline src/services.js]: since it's by amp-form only, should it stay in amp-form?

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 12]
```diff
@@ -249,11 +268,61 @@ export class LinkerManager {
...
+      'type': 'hidden',
+      'name': linkerName,
+      'value': this.resolvedLinkers_[linkerName],
+    });
 
-    el.href = addParamToUrl(href, name, this.resolvedLinkers_[name]);
+    const inputEl = createElementWithAttributes(
+        /** @type {!Document} */ (form.ownerDocument),
+        'input', attrs);
+    form.appendChild(inputEl);
```
Comment (lannka) [inline extensions/amp-analytics/0.1/linker-manager.js]: as discussed offline, we will need to treat method="GET" and method="POST" differently.

if method= "POST", we need to append the params to action URL, instead of hidden input.

File: extensions/amp-analytics/0.1/linker-manager.js [hunk 13]
```diff
@@ -103,7 +109,8 @@ export class LinkerManager {
           this.handleAnchorMutation_.bind(this), Priority.ANALYTICS_LINKER);
     }
 
-    return Promise.all(this.allLinkerPromises_);
+    return Promise.all(this.allLinkerPromises_)
+        .then(this.enableFormSupport_.bind(this));
```
Comment (lannka) [inline extensions/amp-analytics/0.1/linker-manager.js]: let's introduce a feature flag for amp-form support here (so we have full control over the launch).

File: src/services.js (hunk 11 above)
Comment (calebcordry) [reply src/services.js]: It is still a doc level service used by potentially many amp-forms, I though serviced was the preferred method for that?

File: extensions/amp-analytics/0.1/linker-manager.js (hunk 13 above)
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: Done.

File: extensions/amp-analytics/0.1/linker-manager.js (hunk 12 above)
Comment (calebcordry) [reply extensions/amp-analytics/0.1/linker-manager.js]: I will introduce in a follow up PR.
